### ISSUE TITLE
  Connectivity within Nodes

### ISSUE THREAD

**Issue by VishalWaman** [ipfs/go-ipfs#5333]
Hi,

Can you please guide us on below usecase ?

We are working on P2P network implementation over go-ipfs 
We have our own bootstrap nodes for creating network
We need to have seamless discovery +  DHT + Ping and pubsub within them 

Below are challenges we are experiencing : 
1. We are able to  get list of all DHT providers for specific CID (content id)

2. When we try to ping particular node using ID ping fails 
(Ping error: dial attempt failed: <peer.ID cwB1fj> --> 
<peer.ID cncwpp> dial attempt failed: dial tcp4 173.18.0.1:4001: i/o timeout)
 
 and 
 
Error: connect QmWmfowSaHDSzHoeDXAq1jWioc2vSMrr9ZoncK3VqTBDRL failure: dial attempt
 failed: <peer.ID ay2YAo> --> <peer.ID WmfowS> dial attempt failed: dial tcp4
 53.20.142.177:55783: connectex: No connection could be made because the target
 machine actively refused it.

Some Notes:
We are using go-ipfs latest version
We are behind NAT this case was tried on Ec2 with all open ports on provider side

**Comment by Stebalien** [comment]
This looks like https://github.com/libp2p/go-libp2p/issues/383. Dialing `/ip4/54.227.142.187/tcp/4001/ipfs/QmWmfowSaHDSzHoeDXAq1jWioc2vSMrr9ZoncK3VqTBDRL` works but this address isn't being announced:

```
/ip6/::1/tcp/4001
/ip4/172.31.83.102/tcp/4001
/ip6/2001:0:4137:9e76:c5c:bf2:c91c:7144/tcp/4001
/ip4/54.227.142.187/tcp/62719
/ip4/127.0.0.1/tcp/4001
```

Can you confirm?

**Comment by vaibhav6233m** [comment]
Hi @Stebalien, I was working with @VishalWaman  when we were testing this. We are building a desktop application based on ElectronJS and using the `ipfsd-ctl` library to spawn `go-ipfs` daemons.

> Can you confirm? Try connecting directly with ipfs swarm connect /ip4/54.227.142.187/tcp/4001/ipfs/QmWmfowSaHDSzHoeDXAq1jWioc2vSMrr9ZoncK3VqTBDRL

Connecting directly to the node gives success and after the Ping using the peerID also works. 
`ipfs ping QmWmfowSaHDSzHoeDXAq1jWioc2vSMrr9ZoncK3VqTBDRL`.

I also noticed that `ipfs dht findpeer QmWmfowSaHDSzHoeDXAq1jWioc2vSMrr9ZoncK3VqTBDRL` did not include this address `/ip4/54.227.142.187/tcp/4001/ipfs/QmWmfowSaHDSzHoeDXAq1jWioc2vSMrr9ZoncK3VqTBDRL`.

Is there any way to find the correct multiaddresses to the given peerID using `ipfsd-ctl` library.?

Basically we were getting the list of providers for a particular content hash and then we are using the ping api to check if the node is alive. Then using the pubsub model, subscribe to a topic publish a message to which the remote node send a response back with some info.
You can check the following code snippet.
[https://gist.github.com/vaibhav6233m/58b5ca82719c5df037327adcfc126136](url)

**Comment by Stebalien** [comment]
So, the issue here is that your node *knows* that it's external IP is `54.227.142.187` because it dialed a few nodes and those nodes all told your node "you connected to us from `54.227.142.187:62719`". That's why it's announcing this to the network (at least that's what I think is happening).

Unfortunately, it looks like your NAT isn't mapping symmetrically. That is, it's mapping outgoing connections from port `4001` on your machine to port `62719` on the outside of the NAT but it's *not* mapping incoming connections on port `62719` (external) to `4001` on your machine.

@vyzo is working on something called autonat (https://github.com/libp2p/go-libp2p-autonat) that should help with this. Basically, we'll be able to connect to some service and *guess* what our external addresses are. The "autonat" service can then try connecting back to us at those addresses, telling us which ones work and which ones don't. That'll allow us to "guess" that port 4001 *may* work and then test it with the autonat service.

**Comment by Stebalien** [comment]
I'm going to close this in favor of https://github.com/libp2p/go-libp2p/issues/383 to avoid partitioning the conversation.
